bson/codec_options.py

```diff
@@ -33,29 +33,58 @@ def _raw_document_class(document_class):
     return marker == _RAW_BSON_DOCUMENT_MARKER
 
 
+class TypeCodecBase(object):
+    @property
+    def python_type(self):
+        """The Python type to be converted into something serializable."""
+        raise NotImplementedError
+
+    @property
+    def bson_type(self):
+        """The BSON type to be converted into our own type."""
+        raise NotImplementedError
+
+    def transform_bson(self, value):
+        """Convert the given BSON value into our own type."""
+        raise NotImplementedError
+
+    def transform_python(self, value):
+        """Convert the given Python object into something serializable."""
+        raise NotImplementedError
+
+
 class TypeRegistry(object):
-    def __init__(self, encoder_map=None, decoder_map=None):
-        self._encoder_map = (encoder_map or {}).copy()
-        self._decoder_map = (decoder_map or {}).copy()
+    def __init__(self, *args):
```

Comment:
Why `*args` instead of a single argument which is a list or tuple or `TypeCodecBase`'s ? Like
```
    def __init__(self, type_codecs):
```
And if we keep it as `*args` we should still change the name to `*type_codecs` or something else more informative.

Reply:
I liked the idea of being able to initialize this with an arbitrary number of `TypeCodecs` directly rather than having to first create a list/tuple and then pass it to `__init__`. Another reason I went with this was that I don't see what else we might want to pass to this constructor in the future since the type codecs are all self-describing. That said, we can start using a list/tuple if you don't find my reasoning to be satisfactory.
I take your point about the variable name and will change it.

Reply:
I suspect we might add an extra argument to the registry for https://jira.mongodb.org/browse/PYTHON-1731 (something like `TypeRegistry(fallback_encoder=lambda x: repr(x))`) but there could be a better place to add it. Either way let's cross that bridge when we come to it. Keeping it as is in this PR SGTM.

Reply:
Ok! Leaving it be for now.